### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by urjitbhatia**
First pass at #5 
This implementation does not do live persistence but can trigger/restore from a snapshot.

SIGUSR1 triggers a disk persistence event followed by a graceful shutdown.

### CODE REVIEW

**Comment by tommysullivan** [comment]
18,000 lines change! i guess a good bulk are dependencies, do those have to be checked in tho? will have a look this evening. thanx

File: pkg/persistence/journalPersister.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	r := journal.NewReader(f, nil, false, true)
+	if err != nil {
+		err = errors.Wrap(err, "Failed to open peristence journal reader")
+		logrus.Errorf("JournalPersister:Recover %s", err)
+		return nil, err
+	}
+
+	logrus.Info("JournalPersister:Recover streaming items for recovery")
+	go func() {
```
[the remark this answers is not in the document]
Comment (bttmly) [reply pkg/persistence/journalPersister.go]: I guess it uses less memory that way but then its harder to deal with mid-stream errors. You could return `chan []byte, chan error` here I guess

File: pkg/protocol/beanProto.go [hunk 2]
```diff
@@ -67,21 +59,36 @@ type Connection struct {
 }
 
 // NewYaadServer returns a pointer to a new yaad server
-func NewYaadServer(stub bool) *Server {
+func NewYaadServer(stub bool, opts *goyaad.HubOpts) *Server {
```
[the remark this answers is not in the document]
Comment (urjitbhatia) [reply pkg/protocol/beanProto.go]: It gets consumed at a level higher than the Hub itself.
This interface will get cleaned in a bit when I implement the grpc server.

File: pkg/goyaad/job.go [hunk 3]
```diff
@@ -86,3 +92,79 @@ func (j *Job) AsBound(spokeSpan time.Duration) SpokeBound {
 func (j *Job) AsPriorityItem() *Item {
 	return &Item{index: 0, priority: j.triggerAt, value: j}
 }
+
+// GobEncode encodes a job into a binary buffer
+func (j *Job) GobEncode() (data []byte, err error) {
```
[the remark this answers is not in the document]
Comment (urjitbhatia) [reply pkg/goyaad/job.go]: What do you mean dynamic? It is not dynamic at runtime.

If you mean adding new fields: I don't see another way of doing a custom serialization. If this was json and someone could similarly add a new field and not tag it with proper json tags etc.

File: pkg/persistence/journalPersister.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (urjitbhatia) [reply pkg/persistence/journalPersister.go]: @Sticksman recovery is not in the background per-se. The thing calling recover will/should wait for the recovery stream to close.
This does however dedicate a `thread` to read from disk and not have it bogged down by how the rest of the code consumes that data.

@bttmly yeah the chan error could be done but I was thinking that there isn't much you can do with it. If you are restoring say 4 million jobs and 50k are corrupt somehow, there isn't a lot of use of that err chan.
There should probably be a somewhat better error stats reporting at this spot I guess.

**Pull request merged**